File: Graphs/200. Number of islands/solution.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int count=0;
        int maxIsland=0;
        int n=grid.size();
        int m=grid[0].size();
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j] == '1')
                {
                    count++;
                    dfs(i,j,n,m,grid);
                }
            }
        }
        return count;
    }
    void dfs(int i,int j,int n,int m,vector<vector<char>>&grid)
    {
        int x4[]={-1,0,1,0};
        int y4[]={0,1,0,-1};
        if(!(i>=0&&i<n&&j>=0&&j<m))
        {
            return;
        }
        if(grid[i][j]=='0')
        {
            return;
        }
        grid[i][j]='0';
        for(int d=0;d<4;d++)
        {
            int ni=i+x4[d];
            int nj=j+y4[d];
            if(ni>=0&&ni<n&&nj>=0&&nj<m&&grid[ni][nj] == '1')
            {
                dfs(ni,nj,n,m,grid);
            }
        }
        return;
        
    }
};

File: Graphs/200. Number of islands/solution.py
class Solution:
    def __init__(self):
        self.xdir = [1,-1,0,0]
        self.ydir = [0,0,1,-1]
    def dfs(self, i,j, grid):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == "0":
            return
        grid[i][j] = "0"
        for dir in range(4):
            ni = i + self.xdir[dir]
            nj = j + self.ydir[dir]
            if ni >= 0 and ni < len(grid) and nj >= 0 and nj < len(grid[0]) and grid[ni][nj] == "1":
                self.dfs(ni, nj,grid)
        return

    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    count += 1
                    self.dfs(i,j,grid)
        return count
